Fix items not being put into chests.

uecasm opened the pull request:
Closes #383

# Changes proposed in this pull request:
- Fixes chests being placed empty.

Review please

Removing that redundant bit of code (the return value is ignored, the only thing it does is to throw an exception and return because it was given the wrong world) fixes everything.  The chest is populated as expected.  Double chests also work as expected and don't duplicate anything, no matter what rotation is used.  The only thing that continues to not work is mirroring a double chest -- but even then it has the right items, it just doesn't connect properly (which is not new).

uecasm committed:
Fix items not being put into chests.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ldtteam/Structurize?pullRequest=384) <br/>All committers have signed the CLA.

Raycoms reviewed (changes requested):
No, this is here on purpose, if you remove this, players can infinitely dupe items having the builder build chests with content in them.

uecasm commented:
Why?  The builder should demand the items that are in the inventory.  But that would have to happen elsewhere, because this bit of code can't possibly report that anyway.

uecasm commented:
`getRequiredItems` is an entirely separate code path.

Raycoms commented:
> Why? The builder should demand the items that are in the inventory. But that would have to happen elsewhere, because this bit of code can't possibly report that anyway.

The problem is this:
- In the same handler a bit below this code there is a method to query the items. Since chests suck, the items return empty.
- As such the builder won't request any items.
- As such the builder will reach this stage.
- As such, we have to detect in this stage if the items could be queried or not, and if not return an empty inventory.

uecasm commented:
Perhaps then this should only happen if it's a builder operation rather than a creative paste?  Because I can see this being a problem for raid ship chests too as it stands...

Raycoms commented:
> Perhaps then this should only happen if it's a builder operation rather than a creative paste? Because I can see this being a problem for raid ship chests too as it stands...

The solution is simply to make a propper chest handler that is only called for those vanilla containers and that AT the inventory or read the inventory from NBT such that no crash happens.

uecasm committed:
Fix chests differently.

uecasm commented:
Ok, there, that should work.  I haven't tested this with a builder as yet (but I might try in the morning if nobody beats me to it) but I'm confident it will do the job.

uecasm commented:
FWIW, this `loadAllItems` call appears to be used for all of the following in vanilla:
* multiple kinds of furnaces
* barrels
* brewing stands
* campfires
* chests
* dispensers
* hoppers
* shulker boxes

The "since chests suck" code appears to be entirely added by Forge, incidentally (not present in vanilla), and as far as I can see does only affect chests and not any of the other above blocks.  Having said that, it's ultimately still a bug in Structurize (not in Forge) where `getItemStacksOfTileEntity` is passing the wrong World and BlockState to `readTileEntity` during placement (it's passing the target world and block therein, not the blueprint); it's simply a happy accident that most TEs don't seem to care about that.  (This PR is still an improvement, but it would be good to fix the rest of it later.)

Raycoms commented:
> * dispensers

The thing is, that it's impossible to handle double chests otherwise. Even if we create a simulated world just for the placement of chests. We would still get duplicate items from it. As that, on purpose I don't give it the world such that it cannot get the chest. I rather have an empty inventory and place an empty inventory than allowing to dupe items.

Nightenom commented:
Solution to this would be proper multiblock handler imho
With the current state single block nbt parsing is the best thing I see

Raycoms commented:
In the "canHandle" method, we should be able to check if the tileentitydata contains regular inventory data, no?

Nightenom commented:
> In the "canHandle" method, we should be able to check if the tileentitydata contains regular inventory data, no?

Yeah, surely, but remember that this check should be as fast as possible

Raycoms commented:
if container && tenbt.containsx should match most use cases

uecasm commented:
Sure.  Want me to make that change?  Even though with only vanilla blocks it wouldn't actually be needed?  (The other blocks with compatible NBT data aren't multiblocks.)

Raycoms commented:
Sure

On Sun, Jun 6, 2021, 01:47 Gavin Lambert ***@***.***> wrote:

> Sure. Want me to make that change?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/ldtteam/Structurize/pull/384#issuecomment-855336741>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABRD3S4MCGZPOEGKCCGUELTTRL4VTANCNFSM46AX4MTA>
> .
>

uecasm committed:
Fix chests a third way.

This has an API change that will require a trivial change to MineColonies as well.  But it's better.

Raycoms reviewed (approved):
I like this

Nightenom commented on src/main/java/com/ldtteam/structurize/placement/handlers/placement/PlacementHandlers.java:
```diff
@@ -689,14 +689,14 @@ public ActionProcessingResult handle(
     }
 
-    public static class ChestPlacementHandler implements IPlacementHandler
+    public static class ContainerPlacementHandler implements IPlacementHandler
```
Please undo the class rename, as it doesn't seem necessary

uecasm replied:
It handles ContainerBlock, and not just ChestBlock.  The rename seems entirely necessary.

Raycoms replied:
I agree

Nightenom reviewed (commented):
Minor nitpick
